packages/query-core/src/timeoutManager.ts

```diff
@@ -0,0 +1,174 @@
+ */
+export type TimeoutCallback = (_: void) => void
+
+/**
+ * Wrapping `setTimeout` is awkward from a typing perspective because platform
+ * typings may extend the return type of `setTimeout`. For example, NodeJS
+ * typings add `NodeJS.Timeout`; but a non-default `timeoutManager` may not be
+ * able to return such a type.
+ *
+ * Still, we can downlevel `NodeJS.Timeout` to `number` as it implements
+ * Symbol.toPrimitive.
+ */
+export type TimeoutProviderId = number | { [Symbol.toPrimitive]: () => number }
+
+/**
+ * Backend for timer functions.
+ */
+export type TimeoutProvider = {
+  /** Used in error messages. */
+  readonly name: string
+
+  readonly setTimeout: (
+    callback: TimeoutCallback,
+    delay: number,
+  ) => TimeoutProviderId
+  readonly clearTimeout: (timeoutId: number | undefined) => void
+
+  readonly setInterval: (
+    callback: TimeoutCallback,
+    delay: number,
+  ) => TimeoutProviderId
+  readonly clearInterval: (intervalId: number | undefined) => void
+}
+
+export const defaultTimeoutProvider: TimeoutProvider = {
+  name: 'default',
+
+  setTimeout: (callback, delay) => setTimeout(callback, delay),
+  clearTimeout: (timeoutId) => clearTimeout(timeoutId),
+
+  setInterval: (callback, delay) => setInterval(callback, delay),
+  clearInterval: (intervalId) => clearInterval(intervalId),
+}
+
+/** Timeout ID returned by {@link TimeoutManager} */
+export type ManagedTimerId = number
+
+/**
+ * Allows customization of how timeouts are created.
+ *
+ * @tanstack/query-core makes liberal use of timeouts to implement `staleTime`
+ * and `gcTime`. The default TimeoutManager provider uses the platform's global
+ * `setTimeout` implementation, which is known to have scalability issues with
+ * thousands of timeouts on the event loop.
+ *
+ * If you hit this limitation, consider providing a custom TimeoutProvider that
+ * coalesces timeouts.
+ */
+export class TimeoutManager implements Omit<TimeoutProvider, 'name'> {
+  #provider: TimeoutProvider = defaultTimeoutProvider
+  #providerCalled = false
+
+  setTimeoutProvider(provider: TimeoutProvider): void {
+    if (provider === this.#provider) {
+      return
+    }
+
+    if (this.#providerCalled) {
+      // After changing providers, `clearTimeout` will not work as expected for
+      // timeouts from the previous provider.
+      //
+      // Since they may allocate the same timeout ID, clearTimeout may cancel an
+      // arbitrary different timeout, or unexpected no-op.
+      //
+      // We could protect against this by mixing the timeout ID bits
+      // deterministically with some per-provider bits.
+      //
+      // We could internally queue `setTimeout` calls to `TimeoutManager` until
+      // some API call to set the initial provider.
+      console.warn(
+        `[timeoutManager]: Switching to ${provider.name} provider after calls to ${this.#provider.name} provider might result in unexpected behavior.`,
+      )
+    }
```

Comment:
let’s please put this behind an env check and do a `console.error` then, like e.g. here:

https://github.com/TanStack/query/blob/379670dd0a004df872da8494396d30eadcd71d19/packages/query-core/src/query.ts#L539-L543

Reply:
will do 👍🏻. (i took the pattern from searching `console.warn` in the codebase, found one in QueryObserver that i copied)